ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java

```diff
@@ -885,6 +885,141 @@ public void testExpiryAfterMetadataExchange() throws Exception {
+    // the transformer is on local host, which should convert any A1 to A2 (both id and message)
+    Transformer transformer = new BlobIdTransformer(storeKeyFactory, storeKeyConverter);
+
+    Pair<Map<DataNodeId, List<RemoteReplicaInfo>>, ReplicaThread> replicasAndThread =
+        getRemoteReplicasAndReplicaThread(10, clusterMap, localHost, remoteHost, storeKeyConverter, transformer, null);
+    Map<DataNodeId, List<RemoteReplicaInfo>> replicasToReplicate = replicasAndThread.getFirst();
+    ReplicaThread replicaThread = replicasAndThread.getSecond();
+
+    List<RemoteReplicaInfo> singleReplicaList = new ArrayList<>();
+    for (RemoteReplicaInfo replicaInfo : replicasToReplicate.get(remoteHost.dataNodeId)) {
+      if (replicaInfo.getReplicaId().getPartitionId() == partitionId) {
+        singleReplicaList.add(replicaInfo);
+        break;
+      }
+    }
+
+    for (Map.Entry<String, String> caseAndResult : testCasesAndExpectedResults.entrySet()) {
+      // Set up different combinations of PUT, DELETE messages on remote host
+      String[] messages = caseAndResult.getKey().split(" ");
+      for (String message : messages) {
+        switch (message) {
+          case "A1P":
+            addPutMessagesToReplicasOfPartition(Collections.singletonList(blobIdV2),
+                Collections.singletonList(remoteHost));
+            break;
+          case "A1D":
+            addDeleteMessagesToReplicasOfPartition(partitionId, blobIdV2, Collections.singletonList(remoteHost));
+            break;
+          case "A2P":
+            addPutMessagesToReplicasOfPartition(Collections.singletonList(blobIdV5),
+                Collections.singletonList(remoteHost));
+            break;
+          case "A2D":
+            addDeleteMessagesToReplicasOfPartition(partitionId, blobIdV5, Collections.singletonList(remoteHost));
+            break;
+        }
+      }
+      // Do the replica metadata exchange.
+      List<ReplicaThread.ExchangeMetadataResponse> responses =
+          replicaThread.exchangeMetadata(new MockConnection(remoteHost, 10, remoteConversionMap), singleReplicaList);
+      // Do Get request to fix missing keys
+      replicaThread.fixMissingStoreKeys(new MockConnection(remoteHost, 10, remoteConversionMap), singleReplicaList,
+          responses);
+
+      // Verify
+      String resultStr = caseAndResult.getValue();
+      String[] expectedResults = resultStr.equals("") ? new String[0] : resultStr.split(" ");
+      int size = localHost.infosByPartition.get(partitionId).size();
+      assertEquals("Mismatch in number of messages on local host after replication", expectedResults.length, size);
+      for (int i = 0; i < size; ++i) {
+        switch (expectedResults[i]) {
+          case "A1P":
+            assertEquals("Mismatch in blodId on local host after replication", blobIdV2.toString(),
+                localHost.infosByPartition.get(partitionId).get(i).getStoreKey().toString());
```

Comment:
`localHost.infosByPartition.get(partitionId).get(i).getStoreKey().toString()` and `localHost.infosByPartition.get(partitionId).get(i).isDeleted()` are used through out these cases, can you assign them to variables at the beginning of the `for` loop?

Reply:
Changed. Thanks for suggestion.